project3/proto/pub_sub_messages.proto

```diff
@@ -0,0 +1,57 @@
+syntax = "proto3";
+
+package pub_sub_messages;
+
+/// Request sent by a participant to register with the coordinator.
+message Register {
+}
+
+/// Request sent by the coordinator upon successful registration.
+message RegistrationResponse {
+  /// Indicates the ID of this participant.
+  uint32 participant_id = 1;
+}
+
+/// Message sent by a participant to deregister with the coordinator.
+message Deregister {
+}
+
+/// Message sent by a participant to indicate that it is going offline.
+message Disconnect {
+}
+
+/// Message sent by a participant to indicate that it is coming back online.
+message Reconnect {
+}
+
+/// Message sent by a participant that will be sent out to the whole group.
+message SendMulticast {
+  /// The actual message contents.
+  string message = 1;
+}
+
+/**
+ * @brief Message sent by a coordinator to participants containing the data for a
+ * multicast message that has been received.
+ */
+message ForwardMulticast {
+  /// The actual message contents.
+  string message = 1;
+  /// The ID of the participant that this message originated from.
+  uint32 origin_id = 2;
+}
+
+/**
+ * @brief Encapsulates all messages that can be received by the coordinator. This is
+ * so we can simply parse this message type from the network and worry later about
+ * the actual message type.
+ */
+message CoordinatorMessage {
```

Comment:
My only concern with this is that it encapsulates messages that the coordinator will be receiving, but the name implies it contains messages the coordinator will be sending. Should this be instead generalized for both, or possibly renamed to reflect project 1 and 2 structure?

Reply:
I actually don't like the project1/2 structure as much. In practice, you really only need to do this encapsulation for messages that are being received, not send, hence I think it's better to name these encapsulation messages after the receiving end. I know that would certainly make more sense to me if I was skimming the coordinator code and came across a message of this type.

Reply:
Yeah, that does actually make more sense in that case.